app/Main.hs

```diff
@@ -36,33 +41,70 @@ app imageSetRef = do
 
+getRowDivs :: StringLike.StringLike str => Scalpel.Scraper str [str]
+getRowDivs =
+  Scalpel.chroots ("div" Scalpel.@: ["role" Scalpel.@= "row"]) $ do
+    contents <- Scalpel.html Scalpel.anySelector
+    return contents
+
+altTextAndImages :: Scalpel.Scraper String [String]
+altTextAndImages =
+    -- 1. First narrow the current context to each img tag.
+   Scalpel.chroots "img" $ do
+        -- 2. Use Any to access all the relevant content from the the currently
+        -- selected img tag.
+        srcUrl  <- Scalpel.attr "src" Scalpel.anySelector
+        Control.guard ("blob:" `T.isPrefixOf` (StringLike.fromString srcUrl))
+        -- 3. Combine the retrieved content into the desired final result.
+        return (srcUrl)
+
+cleanUp :: Maybe [String] -> Bool
+cleanUp (Just []) = False
+cleanUp _ = True
+
 extractImages :: IORef.IORef (Set.Set T.Text) -> BS.ByteString -> IO BS.ByteString
 extractImages imageSetRef bs = do
     let req = A.decode bs :: Maybe ImageRequest  -- Use bs directly
     case req of
         Just r -> do
-            -- MIO.liftIO $ putStrLn $ "a 2. Decoded request: " ++ show r  -- Debug print
-            let tags = TS.parseTags $ row r
-            MIO.liftIO $ putStrLn $ "a 3. Parsed tags: " ++ show tags  -- Debug print
-            let imgSrcs = Set.fromList [srcValue | TS.TagOpen "img" attrs <- tags, ("src", srcValue) <- attrs, "blob:" `T.isPrefixOf` srcValue]
-            -- MIO.liftIO $ putStrLn $ "a 4. Image sources: " ++ show imgSrcs  -- Debug print
+            let htmlContent = T.unpack $ content r :: String
+            -- putStrLn $ "1. htmlContent: " ++ show htmlContent  -- Debug print
+
+            -- ------NEW CODE
+            -- htmlContent' <- readFromFile "app/test_file.txt" -- get data from testing file, not extension
+            -- putStrLn $ "1. htmlContent: " ++ show htmlContent'  -- Debug print
+            let divs' = Scalpel.scrapeStringLike htmlContent getRowDivs :: Maybe [String]
+            -- putStrLn $ "2. divs': " ++ show divs'  -- Debug print
+            let x = M.fromMaybe [] divs'
+            -- putStrLn $ "3. x: " ++ (L.intercalate "\n  x: " . map show $ x)  -- Debug print
+            let sources = map (\y -> Scalpel.scrapeStringLike y altTextAndImages) x
+            -- putStrLn $ "4. sources: " ++ (L.intercalate "\n  source: " . map show $ sources)  -- Debug print
+            let blobs = M.catMaybes $ filter cleanUp sources
+            -- putStrLn $ "5. blobs: " ++ (L.intercalate "\n  blob: " . map show $ blobs)  -- Debug print
+            -- ------NEW CODE ENDS
+
             alreadyStored <- IORef.readIORef imageSetRef
-            -- MIO.liftIO $ putStrLn $ "a 5. Already stored: " ++ show alreadyStored  -- Debug print
-            let newSrcs = Set.difference imgSrcs alreadyStored
+            let newSrcs = Set.difference (Set.fromList (map T.pack $ concat blobs)) alreadyStored
             IORef.modifyIORef imageSetRef (`Set.union` newSrcs)
```

Comment:
Should this line be moved down so it only happens if `newSrcs` is not empty?

Reply:
yes, you are right!